wallet: remove dead code in legacy wallet migration

rkrux opened the pull request:
A discussion on a previous [PR 32481](https://github.com/bitcoin/bitcoin/pull/32481#discussion_r2145152084) related to legacy wallet dead
code removal made me realize that checking if the legacy
wallet was loaded prior to the start of the migration is not
required ever since legacy wallets can't be loaded in the first
place. I also verified that the `load_on_start` persistent
setting can also not cause the legacy wallets to be loaded, which
further makes the case for removal of the above mentioned checks
during migration.
The current test coverage also shows these lines uncovered.

rkrux committed:
wallet: remove dead code in legacy wallet migration

A discussion on a previous PR 32481 related to legacy wallet dead
code removal made me realize that checking if the legacy
wallet was loaded prior to the start of the migration is not
required ever since legacy wallets can't be loaded in the first
place. I also verified that the `load_on_start` persistent
setting can also not cause the legacy wallets to be loaded, which
further makes the case for removal of the above mentioned checks
during migration.
The current test coverage also shows these lines uncovered.

rkrux commented:
> [previous releases, depends DEBUG](https://github.com/bitcoin/bitcoin/pull/32758/checks?check_run_id=44158542530)Failing after 4m — Task Summary

This seems unrelated as the error is `0curl: (6) Could not resolve host: github.com` in [the CI run](https://github.com/bitcoin/bitcoin/pull/32758/checks?check_run_id=44158542530).

pablomartin4btc reviewed (commented):
Code review ACK

I've left some comments.

I think now the `reload_wallet` helper could be moved into the `if (success) {` condition:
https://github.com/bitcoin/bitcoin/blob/67845e0411241d83c770fafa933b11cf6b8b27b8/src/wallet/wallet.cpp#L4202-L4209

rkrux committed:
wallet: replace `reload_wallet` with inline functionality

Also, update related comments because a reload is not happening
anymore. It is done because the legacy wallets could not have been
loaded prior to migration, so I don't think a reload is happening
post a successful migration, it's just load IMO.

rkrux commented on src/wallet/wallet.cpp:
```diff
@@ -4344,10 +4326,10 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
         // This check is performed after restoration to avoid an early error before saving the backup.
         bool wallet_reloaded = ptr_wallet != nullptr;
-        assert(was_loaded == wallet_reloaded);
+        assert(!wallet_reloaded);
```
Done

rkrux commented on src/wallet/wallet.cpp:
```diff
@@ -4344,10 +4326,10 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
         fs::remove(temp_backup_location);
 
-        // Verify that there is no dangling wallet: when the wallet wasn't loaded before, expect null.
+        // Verify that the legacy wallet is not loaded.
```
Done

rkrux commented:
> I think now the reload_wallet helper could be moved into the if (success) { condition:

Makes sense, I have done this. Also, ended up doing few related changes in the subsequent commits.

pablomartin4btc reviewed (commented):
re-ACK 0635f107c4e1442ce6b953308723b58044c558df

furszy reviewed (commented):
Concept ACK

I think the second commit would look nicer and simpler after something like e86d71b749c08bde6002b9aa2baee824975a518a (which is part of #31423). We’d be able to inline the function at that point.

rkrux commented:
Thanks for highlighting.
I'll take a look at #31423 soon, it had fallen off my radar for some reason.

achow101 commented:
ACK 0635f107c4e1442ce6b953308723b58044c558df

rkrux commented:
Rebased over master to incorporate changes from #31423.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -4304,15 +4295,24 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
+        // Helper to load all the migrated wallets
+        const auto& load_wallet = [&](std::shared_ptr<CWallet>& to_load) {
+            assert(to_load.use_count() == 1);
+            std::string name = to_load->GetName();
+            to_load.reset();
+            to_load = LoadWallet(context, name, /*load_on_start=*/std::nullopt, options, status, error, warnings);
+            return to_load != nullptr;
+        };
+
+        // Migration successful, load all the migrated wallets.
         // Note: We use a pointer to the shared_ptr to avoid increasing its reference count,
-        // as 'reload_wallet' expects to be the sole owner (use_count == 1).
+        // as 'load_wallet' expects to be the sole owner (use_count == 1).
         for (std::shared_ptr<CWallet>* wallet_ptr : {&local_wallet, &res.watchonly_wallet, &res.solvables_wallet}) {
             if (success && *wallet_ptr) {
                 std::shared_ptr<CWallet>& wallet = *wallet_ptr;
                 // Save db path and reload wallet
                 wallet_dirs.insert(fs::PathFromString(wallet->GetDatabase().Filename()).parent_path());
-                success = reload_wallet(wallet);
+                success = load_wallet(wallet);
```
as it is used only here now, could just inline it?

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -4316,7 +4316,7 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(std::shared_ptr<CWallet>
     }
     if (!success) {
-        // Migration failed, cleanup
+        // Either migration failed or loading of the migrated wallets failed, cleanup either way
```
In 2b8b658308c8fbfd9d41d8eb7b9491b615c43872:
Based on the commit description, this change seems more like you reasoning about the code flow than something that benefits the reader? I'm not sure it's adding much value.

rkrux replied:
Dropped the commit.

furszy reviewed (commented):
code reviewed, looks good. Just left a comment about the last commit.

furszy reviewed (commented):
ACK 150b5c99ca11885ef15d04139d919d734e2c211a
